kernel/cClosure.ml

```diff
@@ -1034,6 +1031,35 @@ module FNativeEntries =
       | FArray (_u,t,_ty) -> t
       | _ -> assert false
 
+    let get_blocked env () e =
```

Comment:
This implementation looks wrong to me. I think `unblock` should specify the second argument as needing to be whnf, then you can just look for FApp of `block`.

Reply:
I pushed a fix for this. I originally tried avoiding changing the code too much, but it is a bit hard to avoid adding a new `fterm` constructor for primitives that are values, like `block e`. What I have now works fine in the kernel (see 22936de55a6c8f8f753fc3ecefcb232868d0431d), but cannot be easily replicated in reduction machines based on `Constr.t`.

Morally, we should add a `Blocked of 'types * 'constr` to the `Constr.kind_of_term` type, but that sounds annoying. A better option would be to instead have a new type in `kernel/cPrimitives.ml` for "primitive type constructors", which would also encompass primitive integers, floats and arrays. So, something like:
```ocaml
type ('constr, 'types, 'sort, 'univs) prim_const =
  | Int of Uint63.t
  | Float of Float64.t
  | Array of 'univs * 'constr array * 'constr * 'types
  | Block of 'types * 'constr
```
And then, we would modify the type `Constr.kind_of_term` in the following way:
```diff
-   | Int       of Uint63.t
-   | Float     of Float64.t
-   | Array     of 'univs * 'constr array * 'constr * 'types
+   | PConst    of ('constr, 'types, 'sort, 'univs) prim_const
```

In general, the code around primitives is quite fragile: on master, primitives without any argument of primitive types lead to an assert failure. I fixed a few things, in the MR, but it is unlikely that this is good enough, especially not in the long run.

Basically, I'm not quite sure how to proceed from here, so I'd appreciate some guidance @SkySkimmer @ppedrot.

Reply:
Also, the idea with this new type I'm proposing, my idea is that primitives should be able to manipulate elements of this type parametrically over the type parameters, and that would all happen in the code for running primitives in `kernel/primred.ml`.